apollo-api/src/commonMain/kotlin/com/apollographql/apollo/api/FileUpload.kt

```diff
@@ -1,19 +1,37 @@
 package com.apollographql.apollo.api
 
-class FileUpload(val mimetype: String, val filePath: String) {
-  override fun equals(other: Any?): Boolean {
-    if (this === other) return true
-    if (other !is FileUpload) return false
+import okio.BufferedSink
 
-    if (mimetype != other.mimetype) return false
-    if (filePath != other.filePath) return false
+/**
+ * A class that represents a file upload in a multipart upload
+ * See https://github.com/jaydenseric/graphql-multipart-request-spec
+ *
+ * This class is heavily inspired by [okhttp3.RequestBody]
+ */
+open class FileUpload(val mimetype: String, val filePath: String? = null) {
+  /**
+   * Returns the number of bytes that will be written to `sink` in a call to [.writeTo],
+   * or -1 if that count is unknown.
+   */
+  open fun contentLength(): Long {
+    return -1
+  }
+
+  /**
+   *  Writes the content of this request to `sink`.
+   */
+  open fun writeTo(sink: BufferedSink) {
+    throw UnsupportedOperationException("ApolloGraphQL: if you're not passing a `filePath` parameter, you must override `FileUpload.writeTo`")
+  }
 
-    return true
+  /**
+   * The fileName to send to the server. Might be null
+   */
+  open fun fileName(): String? {
+    throw UnsupportedOperationException("ApolloGraphQL: if you're not passing a `filePath` parameter, you must override `FileUpload.fileName`")
```

Comment:
Should we just remove `filePath` field and make this class abstract? Also make `filename` to be non null abstrac property?

Reply:
That would be much nicer but also a breaking change which I'd like to avoid if possible. I'll do it in dev-3x but given Android 11 is just out, I think it makes sense to release this faster.